lib/superstore/adapters/jsonb_adapter.rb

```diff
@@ -0,0 +1,210 @@
+          value_update = "jsonb_delete(attribute_store, #{fields_to_postgres_array(nil_attributes.keys)})"
+        end
+
+        statement = "UPDATE #{table} SET attribute_store = #{value_update} WHERE #{primary_key_column} = #{quote(id)}"
+        execute_batchable statement
+      end
+
+      def delete(table, ids)
+        statement = "DELETE FROM #{table} WHERE #{create_ids_where_clause(ids)}"
+
+        execute_batchable statement
+      end
+
+      def execute_batch(statements)
+        stmt = [
+          "BEGIN",
+          statements * ";\n",
+          'COMMIT'
+        ] * ";\n"
+
+        execute stmt
+      end
+
+      def create_table(table_name, options = {})
+        define_jsonb_functions!
+
+        ActiveRecord::Migration.create_table table_name, id: false do |t|
+          t.string :id, null: false
+          t.jsonb :attribute_store, null: false
+        end
+        connection.execute "ALTER TABLE \"#{table_name}\" ADD CONSTRAINT #{table_name}_pkey PRIMARY KEY (id)"
+      end
+
+      def drop_table(table_name)
+        ActiveRecord::Migration.drop_table table_name
+      end
+
+      def create_ids_where_clause(ids)
+        ids = ids.first if ids.is_a?(Array) && ids.one?
+
+        if ids.is_a?(Array)
+          id_list = ids.map { |id| quote(id) }.join(',')
+          "#{primary_key_column} IN (#{id_list})"
+        else
+          "#{primary_key_column} = #{quote(ids)}"
+        end
+      end
+
+      def quote(value)
+        connection.quote(value)
+      end
+
+      def fields_to_postgres_array(fields)
+        quoted_fields = fields.map { |field| quote(field) }.join(',')
+        "ARRAY[#{quoted_fields}]"
+      end
+
+      OJ_OPTIONS = {mode: :compat}
+      def to_quoted_jsonb(data)
+        "#{quote(Oj.dump(data, OJ_OPTIONS))}::JSONB"
+      end
+
+      JSON_FUNCTIONS = {
+        # SELECT jsonb_slice('{"b": 2, "c": 3, "a": 4}', '{b, c}');
+        'jsonb_slice(data jsonb, keys text[])' => %{
+          SELECT json_object_agg(key, value)::jsonb
+          FROM (
+            SELECT * FROM jsonb_each(data)
+          ) t
+          WHERE key =ANY(keys);
+        },
+
+        # SELECT jsonb_merge('{"a": 1}', '{"b": 2, "c": 3, "a": 4}');
+        'jsonb_merge(data jsonb, merge_data jsonb)' => %{
+          SELECT json_object_agg(key, value)::jsonb
+          FROM (
+            WITH to_merge AS (
+              SELECT * FROM jsonb_each(merge_data)
+            )
+            SELECT *
+            FROM jsonb_each(data)
+            WHERE key NOT IN (SELECT key FROM to_merge)
+            UNION ALL
+            SELECT * FROM to_merge
+          ) t;
+        },
+
+        # SELECT jsonb_delete('{"b": 2, "c": 3, "a": 4}', '{b, c}');
+        'jsonb_delete(data jsonb, keys text[])' => %{
+          SELECT json_object_agg(key, value)::jsonb
+          FROM (
+            SELECT * FROM jsonb_each(data)
+            WHERE key <>ALL(keys)
+          ) t;
+        },
+      }
+      def define_jsonb_functions!
```

Comment:
Interesting that you had to define all of these functions yourself to implement the low-level jsonb operations.  I'd agree with the comment that @malept made in Slack.  We should performance test this against HSTORE.  I assume the HSTORE operations are implemented in C code in the HSTORE adapter itself, so it may beat the performance of these functions that are implemented in SQL.